[SYCL] Add non-uniform group classes

Pennycook opened the pull request:
Implements basic functionality for the following group types:
- ballot_group
- cluster_group
- tangle_group
- opportunistic_group

This functionality includes all member functions and type traits.
Support for group functions and algorithms will follow later.

cperkinsintel commented:
will there be tests for this over on llvm-test-suite?

Pennycook commented:
> will there be tests for this over on llvm-test-suite?

Yes - the tests for this basic functionality are in https://github.com/intel/llvm-test-suite/pull/1574.  clang-format is currently complaining about them, but otherwise I think they work.

Pennycook commented:
Converting this to draft to prevent it from being merged.  Trying to implement the group algorithms for `masked_sub_group` raised some interesting questions, and I'd like to discuss them some more before committing to this implementation.

Pennycook commented:
> Hi @Pennycook Now `masked_sub_group` is removed from the extension, can you remind me: are you thinking of using the sub_group_mask class directly (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/sub_group_mask.hpp) as a replacement for `masked_sub_group`?

I don't think it's a direct replacement, but it may be helpful to people working with these sorts of masks.  There are still cases where a `sub_group_mask` can be used to reason about which work-items will belong to which groups, and operating on those masks may be more efficient (e.g. because doing some bit-twiddling is probably lower overhead than shuffling between work-items).  The mask could be used to inform control flow that then uses one of the other group types.


> The branch was in a working state but it is now very old. I can remove some of the dead wood in it and try to merge tip in it and post it here so you can have a look. I'm not sure to what extent it might overlap with what you plan, but it sounds like it may be quite similar.

I'd definitely be interested in seeing that, thanks!  I do think a lot of this will end up being useful for implementing the other group types, which are likely to use masks as an implementation detail.

---

Aside: I plan to re-purpose this PR to add the other new group types, rather than close it and open another one.

JackAKirk commented:
> > Hi @Pennycook Now `masked_sub_group` is removed from the extension, can you remind me: are you thinking of using the sub_group_mask class directly (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/sub_group_mask.hpp) as a replacement for `masked_sub_group`?
> 
> I don't think it's a direct replacement, but it may be helpful to people working with these sorts of masks. There are still cases where a `sub_group_mask` can be used to reason about which work-items will belong to which groups, and operating on those masks may be more efficient (e.g. because doing some bit-twiddling is probably lower overhead than shuffling between work-items). The mask could be used to inform control flow that then uses one of the other group types.
> 

OK. I am starting to remember things better now. I think that in the cuda backend we basically only need the `sub_group_mask` for e.g. "cluster_group" or "masked_sub_group" and in particular only the 32 bit mask since e.g. the position of the calling thread wrt the 32 bit subgroup map can be returned from the cuda builtins I referenced above. Then the ptx builtins require that only the threads that are active in the mask call the builtin (if threads that are not active in the mask call the masked ptx builtins you get UB: see e.g. https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-bar-warp-sync), such as bar.warp.sync (sycl::barrier) and redux.sync (sycl reduction algorithms), so within the implementation we simply do e.g.
```    
if (__nvvm_read_ptx_sreg_lanemask_eq() & Mask) {                           
      TYPE result = __nvvm_redux_sync_##REDUX_OP(x, Mask);                     
      *carry = result;                                                         
      return result;                                                           
    } else {                                                                   
      return *carry;                                                           
    }                                                                          
```
Although additionally for post volta maybe we have to think carefully about ensuring the control flow is converged if this is defined by the spec for certain operations as you point out.
 
So is the main point that the design of this will mostly depend on what other hardware needs, and then Nvidia should be pretty easy to fit into whatever design is required?

> > The branch was in a working state but it is now very old. I can remove some of the dead wood in it and try to merge tip in it and post it here so you can have a look. I'm not sure to what extent it might overlap with what you plan, but it sounds like it may be quite similar.
> 
> I'd definitely be interested in seeing that, thanks! I do think a lot of this will end up being useful for implementing the other group types, which are likely to use masks as an implementation detail.
> 

So actually you can get an idea of how it worked from looking at this test branch that is already available: https://github.com/intel/llvm-test-suite/compare/intel...JackAKirk:llvm-test-suite:group_collectives_mask (Note that I think you can ignore all the "async_copy" tests because as I understand it this function is not included in the extension document as a masked variant in any form and the implementations we made for these will be completely removed).
You can see that we just pass a bare sub_group_mask to the "masked" "cooperative" functions, and that we had different types of functions that returned appropriate masks, I think effectively replacing different "group" types like "masked_sub_group" and "cluster_group". I guess this idea is that sub_group_mask will be wrapped in some group classes in order to satisfy implementation requirements of backends other than cuda?

JackAKirk commented:
> Aside: I plan to re-purpose this PR to add the other new group types, rather than close it and open another one.

Also here is the corresponding implementation to go with the above linked tests: https://github.com/intel/llvm/compare/sycl...JackAKirk:llvm:bitmask-cuda

A lot of this is the async stuff which you can ignore. You can completely ignore the changes made to:

libclc/generic/include/clc/async/common.h
libclc/generic/libspirv/async/async_work_group_strided_copy.inc
libclc/ptx-nvidiacl/libspirv/SOURCES
libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl
libclc/ptx-nvidiacl/libspirv/async/wait_group_events.cl.

So I *think* I just need confirmation of the status of the group definitions in this document (https://github.com/AerialMantis/llvm/blob/560269af36543cd88ecd380205b4721abac15354/sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc), and I can update the bitmask-cuda branch linked above to use these group definitions instead of the sub_group_mask helper functions.
So for example, I think I can replace `item.ext_oneapi_partition_sub_group(partition_size)` with `get_cluster_group<partition_size>(sg)` , then update the group function interfaces that were, e.g.:

*version sub_group_mask*:

```
reduce_over_group((sub)Group sg, sub_group_mask mask, T start, Operator binary_op);
```

with a version that just takes a specialized group

*version group*:

```
reduce_over_group((sub/cluster/...)Group sg, T start, Operator binary_op);
```

?

Although I remember you saying that this interface would be changed in some way, so would *version group* not be correct?

Thanks

Pennycook commented:
> Then the ptx builtins require that only the threads that are active in the mask call the builtin (if threads that are not active in the mask call the masked ptx builtins you get UB: see e.g. https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-bar-warp-sync), such as bar.warp.sync (sycl::barrier) and redux.sync (sycl reduction algorithms)

I'm not 100% sure yet, but I think we might be able to skip the "in mask" check.  The groups are defined such that it's UB if the developer calls one of the algorithms using a group that the work-item doesn't belong to.  We might want to keep the check around for debugging purposes, but if it impacts performance we should consider removing it.

> So is the main point that the design of this will mostly depend on what other hardware needs, and then Nvidia should be pretty easy to fit into whatever design is required?

I'd phrase it slightly differently, but I think we're on the same page.  The group types are being designed to be fairly high-level and portable group types that map to common use-cases.  The NVIDIA intrinsics are quite flexible -- because of the arbitrary mask parameter -- so the NVIDIA implementation of each group type really just boils down to figuring out how to construct the mask.

Is that what you meant?

> I guess your idea is that sub_group_mask will be wrapped in some group classes in order to satisfy implementation requirements of backends other than cuda?

I think the `sub_group_mask` group class is to satisfy the implementation requirements of CUDA, where the intrinsics always require a mask.  The SPIR-V non-uniform instructions (e.g. `OpGroupNonUniformShuffle`) do _not_ require a mask, and the work-items involved in a NonUniform instruction are derived from control flow.

Sorry I don't have my code ready to share.  I'll try to get that uploaded by the end of this week -- it should be easier to see the differences, then.

> So I think I just need confirmation of the status of the group definitions in this document

The latest version is [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc), and it includes the changes I alluded to before.  Specifically, `masked_sub_group` has been removed.  The design of the other group types is fixed (for now :laughing:).  I think we can start implementing them, and then address the renaming once we're done.

>  I think I can replace item.ext_oneapi_partition_sub_group(partition_size) with get_cluster_group<partition_size>(sg) (obviously cluster_group name yet to be decided) , then update the group function interfaces

Yes, exactly.  The code in _version group_ is what we're aiming for.

JackAKirk commented:
> 
> I'm not 100% sure yet, but I think we might be able to skip the "in mask" check. The groups are defined such that it's UB if the developer calls one of the algorithms using a group that the work-item doesn't belong to. We might want to keep the check around for debugging purposes, but if it impacts performance we should consider removing it.
> 
OK I see, makes sense.

> The group types are being designed to be fairly high-level and portable group types that map to common use-cases. The NVIDIA intrinsics are quite flexible -- because of the arbitrary mask parameter -- so the NVIDIA implementation of each group type really just boils down to figuring out how to construct the mask.
> 
> Is that what you meant?

Yeah exactly.
>  
> I think the `sub_group_mask` class is to satisfy the implementation requirements of CUDA, where the intrinsics always require a mask. The SPIR-V non-uniform instructions (e.g. `OpGroupNonUniformShuffle`) do _not_ require a mask, and the work-items involved in a NonUniform instruction are derived from control flow.
> 
> Sorry I don't have my code ready to share. I'll try to get that uploaded by the end of this week -- it should be easier to see the differences, then.
> 
OK, this makes sense to me. Thanks for the explanation.

> Yes, exactly. The code in _version group_ is what we're aiming for.

Great, I think we are on the same page. I can work towards this from the cuda backend side via the bitmask-cuda branch.

Pennycook commented:
@JackAKirk I've just added the implementations of the classes themselves (i.e. without group function or algorithm support), and I consider those ready for review.  As you'll see, they're quite straightforward really.

There are two TODOs that I need your help in addressing before things will work on the CUDA or HIP backends:
- In [get_opportunistic_group](https://github.com/intel/llvm/pull/8202/files#diff-ec5a74c5c68ad33ba665455fde899f0cf62d560ace6f0dca76b3f5bd8a9cd3e9R132): I think all that needs to happen here is to construct a `sub_group_mask` containing the active mask.  I know that's `__activemask()` in CUDA, but I'm not sure what the NVVM equivalent is.
- In [get_tangle_group](https://github.com/intel/llvm/pull/8202/files#diff-95000bfe3cd991d95e5d319476a8f18f5f1d56cbd431d8bb877ae7094ee23502R138): This is where we need some experiments/research.  I don't know what the mask value should be here, only that we need the compiler to help build it.

I'm happy to defer these TODOs until later, if you'd like to address them in a later PR.

I'll work on getting the tests up this afternoon, and share a link here when they're available.

Pennycook commented:
@bader - I might be reading the log incorrectly, but it looks like `check-sycl` passed, but the build failed for some other reason.

JackAKirk commented:
> @JackAKirk I've just added the implementations of the classes themselves (i.e. without group function or algorithm support), and I consider those ready for review. As you'll see, they're quite straightforward really.
> 
> There are two TODOs that I need your help in addressing before things will work on the CUDA or HIP backends:
> 
>     * In [get_opportunistic_group](https://github.com/intel/llvm/pull/8202/files#diff-ec5a74c5c68ad33ba665455fde899f0cf62d560ace6f0dca76b3f5bd8a9cd3e9R132): I think all that needs to happen here is to construct a `sub_group_mask` containing the active mask.  I know that's `__activemask()` in CUDA, but I'm not sure what the NVVM equivalent is.
> 
>     * In [get_tangle_group](https://github.com/intel/llvm/pull/8202/files#diff-95000bfe3cd991d95e5d319476a8f18f5f1d56cbd431d8bb877ae7094ee23502R138): This is where we need some experiments/research.  I don't know what the mask value should be here, only that we need the compiler to help build it.
> 
> 
> I'm happy to defer these TODOs until later, if you'd like to address them in a later PR.
> 
> I'll work on getting the tests up this afternoon, and share a link here when they're available.

For activemask it does not currently have a nvvm builtin upstream. I added it here: https://github.com/intel/llvm/compare/sycl...JackAKirk:llvm:bitmask-cuda and also added `__spirv_GroupActiveItems` which calls the nvvm builtin: I think that adding `__spirv_GroupActiveItems` was probably not a good idea since HIP also does not appear to support a corresponding instruction, and probably we should call `activemask` directly in sycl runtime as you've done here using the `__NVPTX__` macro. However upstream uses `activemask` in a clang builtin via inline asm using volatile for some reason. I need to understand why they did this. But this is just a small detail.

For the corresponding tangle_group `__NVPTX__` case, do you think that we should do a check for sm version here and return a mask with only the single calling thread active if >=sm_70 (volta), and then a mask with all 1's for pre-volta?
That corresponds to what my current understanding of what tangle_group means for Nvidia, when it is constructed from a sub_group.

Pennycook commented:
> For activemask it does not currently have a nvvm builtin upstream. I added it here: [sycl...JackAKirk:llvm:bitmask-cuda](https://github.com/intel/llvm/compare/sycl...JackAKirk:llvm:bitmask-cuda) and also added `__spirv_GroupActiveItems` which calls the nvvm builtin: I think that adding `__spirv_GroupActiveItems` was probably not a good idea since HIP also does not appear to support a corresponding instruction, and probably we should call `activemask` directly in sycl runtime as you've done here using the `__NVPTX__` macro. However upstream uses `activemask` in a clang builtin via inline asm using volatile for some reason. I need to understand why they did this. But this is just a small detail.

Ah, I see.  I agree we should call `activemask` directly.  I think the `__spirv` prefix actually triggers special handling by the SPIR-LLVM translator, too, so we should be careful not to introduce functions with that naming convention that don't correspond to a SPIR-V instruction.

> For the corresponding tangle_group `__NVPTX__` case, do you think that we should do a check for sm version here and return a mask with only the single calling thread active if >=sm_70 (volta), and then a mask with all 1's for pre-volta? That corresponds to what my current understanding of what tangle_group means for Nvidia, when it is constructed from a sub_group.

I hadn't considered checking the SM version here.  That's a good idea.

The mask values you've proposed wouldn't work for `tangle_group`, though.   This concept has tripped a few people up, now, so we should give some thought to how we can explain it better in the extension.  The members of a `tangle_group` are supposed to be the same across all hardware, and the only thing that should impact group membership is the control flow.

I've sketched out an example with some multiple branches below, showing the expected membership of a `tangle_group` at each point:

```c++
// At the start of a kernel, all work-items are active
// All work-items will encounter this call
// mask = 0xffffffff (equivalent to using the sub-group itself)
auto t1 = it.get_tangle_group(sg);

if (sg.get_local_id() < sg.get_local_range() / 2) {
  // Only the lower half of the sub-group will encounter this call
  // mask = 0x0000ffff;
  auto t2 = it.get_tangle_group(sg);

  if (sg.get_local_id() % 2 == 0) {
    // Only the even members will encounter this call
    // NB: The tangle must be a subset of the lower half, because of the previous branch
    // mask = 0x00005555
    auto t3 = it.get_tangle_group(sg);
  }
}
```

For SM < 70, this should be the behavior of `__activemask`.  Any lanes on the same control flow should be implicitly synchronized.  I don't think we can use all 1s, because some of the work-items won't be active.  (But that could be an implementation detail -- if the mask value doesn't really matter pre-Volta, all 1s might work just as well as the active mask).

For SM >= 70, I think the compiler needs to track the control flow and build a stack of masks, as-if you'd run something like:
```c++
// assume this stack is shared across all work-items in the sub-group
// push/pop only called once per sub-group
auto masks = std::stack<uint32_t>();
masks.push(0xffffffff);
auto t1 = /* construct tangle group from masks.top() */

bool b1 = (sg.get_local_id() < sg.get_local_range() / 2);
uint32_t m1 = __ballot_sync(masks.top(), b1); // 0x0000ffff;

masks.push(m1);
if (b1) {
  auto t2 = /* construct tangle group from masks.top() */

  bool b2 = (sg.get_local_id() % 2 == 0);
  uint32_t m2 = __ballot_sync(masks.top(), b2); // 0x00005555;

  masks.push(m2);
  if (b2) {
    auto t3 = /* construct tangle group from masks.top() */
  }
  masks.pop();
  __syncwarp(masks.top()); // wait for all work-items that took b1, including those that didn't take b2
}
masks.pop();
__syncwarp(masks.top()); // wait for all work-items, including those that didn't take b1
```

I _think_ that the transformation above effectively forces SM >= 70 to behave as-if it were SM < 70.  But we would need a dedicated compiler pass to build and maintain this stack of masks, insert calls to `__ballot_sync` and `__syncwarp` where appropriate, and ensure that the stack was visible across function calls.

Does that make sense?

JackAKirk commented:
I've now thought about these new groups more carefully and done some implementing. Apart from the open questions around `tangle_group`, the only one I have concerns for is `ballot_group`. Making a cuda implementation of `ballot_group` requires a mask, and I do not see how such the mask can be constructed either from a ballot_group instance or during its construction. I may be missing something of course; let me discuss `cluster_group` first that may be useful to consider wrt the above.

For `cluster_group` I can construct the mask within `get_cluster_group` by using the parent group: I have a working implementation of this hooked up to the nvidia backend implementation for the `reduce_over_group` taking `cluster_group`. Note that it would not be so straightforward to construct the mask after `cluster_group` were instantiated, say at the point that the mask is required: say when e.g. `reduce_over_group` or `group_barrier` are called in the __NVPTX__ backend, since e.g. there is no information saved currently in these group functions for the position of the group item in the non-uniform group wrt to parent subgroup (which I used to construct the mask); but I imagine it makes sense to construct the mask during the cluster_group construction and introduce a mask member for cluster_group in the nvptx backend using __NVPTX__ anyway (this is what I have done in my test implementation)?

But going back to `ballot_group`, even if it were desirable to construct the mask after the `ballot_group` construction, without some information of how the non-uniform subgroup instance calling thread maps to the parent subgroup, I'm not sure how this would be done. And I think it would be desirable to construct the mask when `ballot_group` is constructed, in which case I don't immediately see a way to do this without passing the complete mask to the `ballot_group` constructor explicitly.

Pennycook commented:
> Yes, I think this makes sense. Do you want us to investigate implementing tangle_group along these lines in the cuda backend now, or do you think it would be OK to introduce tangle_group as an intel extension for the time being?

I think we can implement the other group types, first.  Because all the CUDA groups are expected to be mask-based under the covers, we can get the algorithms working for the other groups before sorting out the required mask values for `tangle_group`.

We shouldn't put things on the back-burner for _too_ long, though; it would be good to make sure that `tangle_group` can be implemented portably before we encourage too many people to use it.

>  Note that it would not be so straightforward to construct the mask after cluster_group were instantiated, say at the point that the mask is required: say when e.g. reduce_over_group or group_barrier are called in the NVPTX backend, since e.g. there is no information saved currently in these group functions for the position of the group item in the non-uniform group wrt to parent subgroup (which I used to construct the mask); but I imagine it makes sense to construct the mask during the cluster_group construction and introduce a mask member for cluster_group in the nvptx backend using NVPTX anyway

In the general case (where a `cluster_group` is constructed from an arbitrary group) I agree with you -- we should just store a mask inside the class upon construction.  For the specific case where a `cluster_group` is constructed from a `sub_group`, though, I think you _can_ reconstruct the mask upon request (e.g. the lane within the cluster is the warp lane ID % `ClusterSize`, and the warp lane ID can be computed from the thread ID registers).

> But going back to ballot_group, even if it were desirable to construct the mask after the ballot_group construction, without some information of how the non-uniform subgroup instance calling thread maps to the parent subgroup, I'm not sure how this would be done.

`ballot_group` accepts a group parameter, which I think is sufficient to address this.  If the group passed in is one of the user-constructed groups, then the group object contains an explicit mask that can be used inside the `ballot_group` constructor.  If the group passed in is one of the fixed topology groups (e.g. sub-group) then the mask is implicit (e.g. it's always `0xffffffff` when constructing a `ballot_group` from `sub_group`):

Pennycook commented:
@cperkinsintel, @intel/llvm-reviewers-runtime: I think this is ready for your review; although there are a small number of CUDA-related TODOs, the majority of the code is backend-independent and I don't expect it to change.

JackAKirk commented:
> 
> I think we can implement the other group types, first. Because all the CUDA groups are expected to be mask-based under the covers, we can get the algorithms working for the other groups before sorting out the required mask values for `tangle_group`.
>
OK makes sense.
> 
> We shouldn't put things on the back-burner for _too_ long, though; it would be good to make sure that `tangle_group` can be implemented portably before we encourage too many people to use it.
> 
OK sure, we can look into how to implement tangle group for cuda now.
> `ballot_group` accepts a group parameter, which I think is sufficient to address this. If the group passed in is one of the user-constructed groups, then the group object contains an explicit mask that can be used inside the `ballot_group` constructor. If the group passed in is one of the fixed topology groups (e.g. sub-group) then the mask is implicit (e.g. it's always `0xffffffff` when constructing a `ballot_group` from `sub_group`).
>
Yes I think I understand and you are right on all accounts. I think it would make sense that in the cuda implementation all of these non-uniform sub-groups construct and store the mask on the group construction, so it only happens once at the beginning. Do you think this is a good idea?

In the example that you have for `ballot_group` in the extension doc, if I comment out the `if` statement will it make the call to the group functions UB according to the extension spec? I am thinking that it would be in the cuda backend if the implementation isn't careful because work-items that are not in the mask will be executing a ptx instruction that states this is UB.

```
auto sg = it.get_sub_group();

// get group representing the subset of the sub-group that will take the branch
auto will_branch = sg.get_local_linear_id() % 2 == 0;
auto inner = sycl::ext::oneapi::experimental::get_ballot_group(sg, will_branch);

//if (will_branch)
{
  // wait for all work-items that took the branch to hit the barrier
  sycl::group_barrier(inner);

  // reduce across subset of outer work-items that took the branch
  float ix = sycl::reduce_over_group(inner, x, plus<>());
}
```

Pennycook commented:
> I think it would make sense that in the cuda implementation all of these non-uniform sub-groups construct and store the mask on the group construction, so it only happens once at the beginning. Do you think this is a good idea?

Yes, I think that makes sense.  As I work through the SPIR-V prototype I'm finding cases where the SPIR-V implementation needs to store a group, too, so there will be some consistency in implementation across backends.

> In the example that you have for `ballot_group` in the extension doc, if I comment out the `if` statement will it make the call to the group functions UB according to the extension spec, or not? Like:
> ...
> But if the intention is that when `ballot_group` is instantiated, in the cuda backend the mask that is created will be different depending upon the predicate, such that there are two masks created, and each executing thread is always assigned a corresponding mask in which that thread is active, then everything will be fine. I guess this is what you wanted anyway.

Great question.  I think the way that `ballot_group` is described today, the second behavior is the correct one.  The specification says that `ballot_group::get_group_range()` always returns 2, and `ballot_group::get_group_id()` is 0 or 1 (corresponding to the predicate).  So, the mental model here is that `get_ballot_group()` doesn't create one group, but rather partitions an existing group into two groups based on the value of the predicate.

The way this works in my prototype right now is:
```c++
  sub_group_mask mask = sycl::ext::oneapi::group_ballot(group, predicate);
  if (predicate) {
    return ballot_group<sycl::sub_group>(mask, predicate); // protected constructor to pass in mask and predicate
  } else {
    return ballot_group<sycl::sub_group>(~mask, predicate); // ditto; note that mask is inverted for the second group
  }
```

JackAKirk commented:
Hi @Pennycook 

Could you comment on the approach here of introducing masked variants of spv builtins defined in tablegen: https://github.com/intel/llvm/compare/sycl...JackAKirk:masked_sub_group_cuda?expand=1#diff-73b1c844b6929ccc37836285e2d87f89ead7074c6693a80e712f06bb54fa5cb2R941

That are then used e.g. here https://github.com/intel/llvm/compare/sycl...JackAKirk:masked_sub_group_cuda?expand=1#diff-b5f3e302dbe3b7f867c0885a4709666fe502ef11043e8954a2883aa78736f458R363 , in order to match the pattern of the implementation for the non-masked group functions. In the cuda/hip side the spirv functions are then implemented in libclc as with the existing group algorithms.
This avoid e.g. __NVPTX__ or __AMDGCN__ or __SPIR__ macro usage, but I want to know what you think about it from the intel __SPIR__ side? Do you think it is a good idea or not?

JackAKirk commented:
> 
> I suppose the question is: do you want to make a spirv extension for these masked variants?

To be specific, based on the existing spirv spec, I think it would make sense to add Masked version as:

```
 __spirv_GroupExclusiveScanMasked
__spirv_GroupInclusiveScanMasked
__spirv_GroupBitwiseOrMasked
__spirv_GroupBitwiseAnyMasked
__spirv_GroupBroadcastMasked
__spirv_ControlBarrierMasked
__spirv_GroupReduceMasked
```

Corresponding to existing spirv functions, motivated by the requirements of the group functions supporting the non-uniform subgroups from https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc
What do you think?

Also I think it would basically be required to add e.g.:

```
spirv::GroupBroadcastMasked
spirv::SubgroupShuffleMasked
spirv::SubgroupShuffleXorMasked
spirv::SubgroupShuffleUpMasked
spirv::SubgroupShuffleDownMasked
spirv::GroupAnyMasked<Group>
spirv::GroupAllMasked<Group>
```

to corresponding non masked functions from spirv.hpp. But as I understand it these guys aren't part of a spirv specification?

Pennycook commented:
Sorry for the delayed response, @JackAKirk. 

I think I would prefer if we didn't add a new SPIR-V extension at this time.  We still don't know exactly when and where the masked variants need to be used, and I suspect we'll still need some `#ifdef NVPTX` stuff anyway -- there are cases where I believe the Intel and AMD backends won't call the masked variant, but the NVIDIA backend will.

For example, I think `tangle_group` can be implemented on Intel using `OpGroupNonUniformReduce` (without a mask), and `cluster_group` can be implemented using `OpGroupNonUniformReduce` with the group operation set to `ClusteredReduce`.

It may also worth noting that NVIDIA have [their own SPIR-V extension](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/NV/SPV_NV_shader_subgroup_partitioned.asciidoc) exposing some of the mask support.  I don't think it has support for everything you'd need for the NVIDIA backend, but it might be worth keeping an eye on it.

> Also I think it would basically be required to add e.g.:
> `spirv::GroupBroadcastMasked`

I think we should do this, and that should be our solution to a common interface here (for now).

I've already got functions like `spirv::NonUniformBroadcast` in my prototype, and I think we should add functions like `spirv::NonUniformMaskedBroadcast` as you suggest.  (Note though that I think we should keep "NonUniform" somewhere in the name, to help people understand which control flow rules they follow).

That way, anywhere that all the backends use the same logic, we can all call the same `spirv::` function.  And if I'm right about requiring `ifdef NVPTX` stuff, we can hide it in those functions instead of polluting the group algorithm implementation.

JackAKirk commented:
> Sorry for the delayed response, @JackAKirk.
> 
> I think I would prefer if we didn't add a new SPIR-V extension at this time. We still don't know exactly when and where the masked variants need to be used, and I suspect we'll still need some `#ifdef NVPTX` stuff anyway -- there are cases where I believe the Intel and AMD backends won't call the masked variant, but the NVIDIA backend will.
> 
OK I see makes sense. Thanks.


> It may also worth noting that NVIDIA have [their own SPIR-V extension](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/NV/SPV_NV_shader_subgroup_partitioned.asciidoc) exposing some of the mask support. I don't think it has support for everything you'd need for the NVIDIA backend, but it might be worth keeping an eye on it.
> 
Thanks I didn't know this.

JackAKirk commented:
This LGTM.

cperkinsintel commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1574

Pennycook commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1574

Pennycook commented:
@cperkinsintel, @bader: The failures didn't seem to be related to any of the changes here (something about a missing path on the CI machines?).  I can't seem to get things to re-run, either; could you please take a look?

Pennycook commented:
Closed and reopened in an attempt to retrigger the checks.

cperkinsintel commented:
@Pennycook   I'm probably telling you something you already know, but on Jenkins/llvm-test-suite it looks like there are failing tests.
 ```Failed Tests (2):
[2023-03-07T16:22:47.263Z]   SYCL :: NonUniformGroups/is_fixed_topology.cpp
[2023-03-07T16:22:47.263Z]   SYCL :: NonUniformGroups/is_user_constructed.cpp```

Pennycook commented:
> @Pennycook I'm probably telling you something you already know, but on Jenkins/llvm-test-suite it looks like there are failing tests.
> 
> ```
> Failed Tests (2):
> [2023-03-07T16:22:47.263Z]   SYCL :: NonUniformGroups/is_fixed_topology.cpp
> [2023-03-07T16:22:47.263Z]   SYCL :: NonUniformGroups/is_user_constructed.cpp```
> ```

@cperkinsintel - I'd actually missed that the tests started running correctly, so thanks for letting me know.  I think I've fixed things, so I'll try and re-run that job.

Pennycook commented:
@cperkinsintel: Things seem to be passing now.

steffenlarsen commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1574

cperkinsintel commented:
@Pennycook  - failure on LevelZero GPU
```
Failed Tests (1):

[2023-03-10T15:18:46.792Z]   SYCL :: NonUniformGroups/tangle_group.cpp
```

Pennycook commented:
> @Pennycook - failure on LevelZero GPU
> 
> ```
> Failed Tests (1):
> 
> [2023-03-10T15:18:46.792Z]   SYCL :: NonUniformGroups/tangle_group.cpp
> ```

Do you know which GPU we test on?  I know this is cliche, but the test passes on my machine!

Pennycook commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1574

Pennycook commented:
@intel/llvm-reviewers-runtime: @steffenlarsen and I were able to implement a workaround for the failing test.  We're going to need to revisit this later on, but I think this can be merged now.

bader commented on sycl/include/sycl/sycl.hpp:
```diff
@@ -69,9 +69,13 @@
+#include <sycl/ext/oneapi/experimental/cluster_group.hpp>
 #include <sycl/ext/oneapi/experimental/cuda/barrier.hpp>
+#include <sycl/ext/oneapi/experimental/opportunistic_group.hpp>
+#include <sycl/ext/oneapi/experimental/tangle_group.hpp>
```
The compile time is not great already. Can we don't add these includes to `sycl.hpp` and ask users to include header files with experimental API explicitly?

Pennycook replied:
I'm not opposed to that, but would it be acceptable to do that in a separate PR?  There are some other experimental headers that I think we'd want to move (e.g. `cuda_barrier`) and some other headers that aren't obviously marked experimental but which implement experimental features (e.g. kernel properties).

I think it could be confusing if these new groups are the only experimental feature requiring users to include extra headers.

If we were to adopt this as a convention for experimental features, I think we'd need to update the documentation for our extensions too. So I'd like to hear @gmlueck's opinion about that.

bader replied:
I suggest we remove all includes of headers with API extensions (not just experimental ones) and leave only headers with standard features API in `sycl.hpp`. I remember discussing this already, but I don't remember if we decided to clarify this in any documentation. On the other hand, as far as I know there are no guidelines to add includes for extension headers to standard `sycl.hpp` header neither.
I'm not sure if we do it easily for existing includes as it might be considered as API breaking change, but at least we can avoid doing it for new headers. Once breaking changes are allowed, we can clean-up `sycl.hpp` file.

> would it be acceptable to do that in a separate PR?

I think so, but, please, make sure doing this before any products based on intel/llvm are released. Otherwise, it's going to be an API breaking change and we usually keep API stable for a long period of time.

gmlueck replied:
It is definitely an API breaking change if we start requiring the user to `#include` a new header in order to use an extension.

I'd like to have a consistent strategy before we start changing things.  I agree with @Pennycook that it will be confusing if only a few extensions require the user to `#include`.  Possible options are:

* Require users to `#include` for experimental extensions but not supported or deprecated ones.
* Require users to `#include` for all extensions.  In this case, we need some plan to manage the transition.  For example, can we keep existing extensions as part of `sycl/sycl.hpp` with a deprecation warning?

Note that some extensions will not be able to be extracted into a separate header.  For example, extensions that add member functions to existing classes don't have a separate header.

What about extensions that add new types *and* also add new member functions to existing classes?  It might be weird to have the extension half supported when the user didn't `#include` the extension's header.

Maybe it would be better to have all the extensions implemented in `sycl/sycl.hpp`, but enable them via some command line option.  The option could predefine some macro which enables the extension code within the headers.

Pennycook replied:
> Maybe it would be better to have all the extensions implemented in sycl/sycl.hpp, but enable them via some command line option. The option could predefine some macro which enables the extension code within the headers.

Since every extension already defines a macro, could we just use those?

I could then update sycl.hpp to:
```c++
#ifdef SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS
#include <sycl/ext/oneapi/experimental/opportunistic_group.hpp>
#include <sycl/ext/oneapi/experimental/tangle_group.hpp>
#endif
```

(Or, alternatively, put that `#ifdef` inside each of the group headers.)

Then we could tell developers that to use experimental extensions they have to enable the macro themselves, whereas the macros for other (non-experimental) extensions are predefined?

> I'm not sure if we do it easily for existing includes as it might be considered as API breaking change, but at least we can avoid doing it for new headers. Once breaking changes are allowed, we can clean-up sycl.hpp file.

If we only make changes for experimental extensions, do breaking changes matter?

bader replied:
Following [Zero-overhead_principle](https://en.cppreference.com/w/cpp/language/Zero-overhead_principle) "You don't pay for what you don't use" C++ standard library API is distributed across multiple headers: https://en.cppreference.com/w/cpp/header. Users include `<optional>` header to import `std::optional` class declaration. 

GNU libstdc++ implementation provides a non-standard `<bits/stdc++.h>` header file, which includes all the headers of C++ standard library. Using this header simplifies the use of C++ standard library functionality but incurs the overhead of parsing all the headers.

The SYCL specification defines a single header for the entire standard library/API. The spec doesn't define whether extensions should be defined in `sycl/sycl.hpp`, it just says that extensions can add new headers, and if they do, they must follow the naming scheme. The more features we add, the higher the overhead on parsing `sycl/sycl.hpp`.

Technically, the DPC++ library already provides separate headers for each class or feature, but we use them only for internal testing and don't promote using them to our users.

I believe that SYCL specification should provide users ability to include reduced set of declarations similar to standard C++ library. Having `sycl/sycl.hpp` including all standard declarations and extensions is fine as long as we have an alternative to avoid "usability overhead".

> Since every extension already defines a macro, could we just use those?

I don't think so. If feature doesn't require compiler support, the macro is usually defined in the header itself ([example](https://github.com/intel/llvm/blob/27654ac69fe26828ef7e5fb4a2a1bdac7dd1f86b/sycl/include/sycl/ext/oneapi/experimental/uniform.hpp#L14-L16)). The macro can be tested after include directive.

> Then we could tell developers that to use experimental extensions they have to enable the macro themselves, whereas the macros for other (non-experimental) extensions are predefined?

The macro is supposed to be set by implementation as it indicates whether extension is enabled by implementation. SYCL standard macro defined by user might be confusing. But we could have some controls to disable support for extensions/feature (e.g. compiler flag and/or `SYCL_EXT_DISABLE_***` macro for use before SYCL headers included).

We should move this discussion to `SYCL-Docs` project as we are not going to solve this problem in this PR, but I wanted to raise this issue to draw the attention to the compile time overhead problem.

Pennycook replied:
I'm 100% behind splitting SYCL out into multiple headers (or C++20 modules) to improve compile-times.  Any resistance I'm showing here is definitely self-serving: I'd like to get these initial class definitions merged so that we can move our attention to things like CUDA support and more complicated functionality (e.g. the algorithms).

You're right that the macro fix I proposed is probably just as confusing as requiring extra includes.  I agree that this is likely a bigger problem that we should discuss elsewhere... I suspect that it will take some time to reach broad consensus about what to do here, so what should I do to move this PR forward?

bader replied:
I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.

Having said that, this is not the most critical issue for this project, and we will need to address this problem for other features anyway. I won't block your patch, if you want to commit it with `sycl.hpp` changes.

gmlueck replied:
> I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
> There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.

This would be a good topic to discuss in our Friday "SYCL Language Discussion" meeting.  Would you like to summarize the current status of experimental extensions and their headers, and then present at the meeting?

bader replied:
> > I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
> > There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.
> 
> This would be a good topic to discuss in our Friday "SYCL Language Discussion" meeting. Would you like to summarize the current status of experimental extensions and their headers, and then present at the meeting?

Unfortunately, I don't track DPC++ development closely enough to feel like I can present a summary of the current status. At the same time, I don't think the current status of experimental extensions and their headers is really important to the main topic of this discussion - compile time overhead. Experimental extensions are brought here as an example of uncontrolled compile time overhead accumulation because this PR implements an experimental extension and there is an easy way to avoid it (my POV).

There are multiple ways to deal with the compile time overhead problem and the SYCL specification can help with making the solution portable across multiple implementations. I will glad to join the discussion at the meeting if you send me invitation.

bader commented on sycl/include/sycl/ext/oneapi/experimental/opportunistic_group.hpp:
```diff
@@ -0,0 +1,148 @@
+  sycl::sub_group sg = sycl::ext::oneapi::this_sub_group();
+  sub_group_mask mask = sycl::ext::oneapi::group_ballot(sg, true);
+  return opportunistic_group(mask);
+#elif defined(__NVPTX__)
```
DPC++ supports AMD GPU targets in addition to these targets.
Is it possible to give a diagnostic on unsupported targets?

Pennycook replied:
Honestly, I'm not sure.  I thought about this, but I was worried that if I emitted something like a `#warning` then it would trigger on every AMD GPU compilation that included the headers (rather than only compilations calling the function).

@steffenlarsen, is there a way to trigger a compile-time or run-time diagnostic when a function is invoked on an unsupported backend?  The only precedent I see for this is throwing a `runtime_error` on the host device.

steffenlarsen replied:
I agree. A warning would likely leak into any AMDGCN/NVPTX compilations. Had this function been template dependent we could have used a static assertion using the template argument. I believe something like

```c++
template <typename T = void, typename = std::enable_if_t<std::is_same_v<T, void>>>
inline opportunistic_group get_opportunistic_group() {
...
#elif defined(__NVPTX__) || defined (__AMDGCN__)
  static_assert(!std::is_same_v<T, void>, "get_opportunistic_group() is not currently supported on the target");
#endif
...
}
```
would work, but I am not sure it's worth the dependence.

gmlueck replied:
When we say "not currently supported" is this something that we expect to fix soon, or is it a long term problem?

I think the `static_assert` solution also does not work in the general case.  Imagine an application that has some kernels that run on Nvidia and some that run on Intel GPUs.  The application use `get_opportunistic_group` in a kernel that will be submitted to an Intel GPU and the host code checks the device type before submitting the kernel.  There are other kernels in the same TU that can run on Nvidia devices.

Wouldn't the `static_assert` trigger in that case also?

We invented the "optional kernel features" concept to handle cases like this.  The diagnostic occurs at runtime if the application mistakenly submits the kernel using `get_opportunistic_group` to an Nvidia device.

Pennycook replied:
> When we say "not currently supported" is this something that we expect to fix soon, or is it a long term problem?

From my perspective, this is a complete unknown right now, and it's why everything is marked experimental.

My aim is to get things running reliably across all the devices and backends, but we need to get implementation experience.  I sketched out `__SPIRV__` and `__NVPTX__` paths here because I understand what the implementations will look like; I need somebody with AMDGCN experience to take a look at things there.

Designing aspects and optional features for these group types right now seems premature to me, because we don't know which group types we'll need to make optional.  If it's not a lot of work for somebody to give us optional features for each group type then I'm not opposed to that, but I don't want to make a lot of extra work for folks if there's a chance we won't need the aspects later on.

bader replied:
I wanted to make sure that when this experimental feature is used on unsupported platforms, users will receive a meaningful error message about it. Actually I think we might already have it. I see that only code under `#if defined(__SPIR__)` has `return` statement, so I expect if `__SPIR__` is not defined the compiler will report an error about missing `return`.
@Pennycook, do you have any tests for this API we can run on NVIDIA/AMD GPUs?

Pennycook replied:
The tests in https://github.com/intel/llvm-test-suite/pull/1574 will work on anything supporting a sub-group size of 32 (once the NVIDIA/AMD paths are implemented).  I've marked those backends as unsupported on NVIDIA/AMD for now because they're unimplemented, but @JackAKirk is working on implementing the missing parts.

bader replied:
@JackAKirk, do you know whether DPC++ with #8202 gives any diagnostic on NVIDIA/AMD GPU for using new classes? I hope the compiler returns an error. Am I right?

John, meaningful diagnostic is nice to have, but should not block commit of experimental API. I suppose at this point all users of this feature know how to use it properly and what to do if it doesn't work as intended. We just need to make sure that when it's ready for widespread use, the behavior of the implementation is user-friendly.

Sorry for the inconvenience I cause by discussing non-critical matters.

steffenlarsen replied:
@gmlueck - I agree, the static assertion would cause problems in cases like the one you mentioned. Having an aspect for this and marking the relevant functions as requiring that aspect is definitely an option.

JackAKirk replied:
For the groups like opportunistic_group that weren't implemented on NVPTX I think my memory is that the tests did not return an error immediately when `get_opportunistic_group` was called, but I think as @steffenlarsen points out this could maybe be fixed with a static assert like: https://github.com/intel/llvm/blob/9596ea0ddd362165cefd419e57cf2b57f5962c6c/sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp#L32
For all types added here except for tangle_group, I think that we resolved the discussions on what the NVPTX implementation should be, so I could push what I implemented for NVPTX after this PR is merged. For AMD I think that AMD gpus do not support independent forward progress and therefore don't have a similar instruction to __activemask from NVPTX, but I think that for AMD get_opportunistic_group can be supported in the essentially same was as it is for __SPIR__ here, by using ballot, the only thing is that amd also needs to support 64 WI subgroups.

Pennycook commented:
@bader - Please merge this as-is.  I've made a note to myself to revisit these questions when we're done with the initial implementation and/or we've reached consensus on the experimental headers issue.

And there's no need to apologize! I hadn't actually considered what warning/error behavior we might want in the final implementation, so this has given me a lot to think about.
